graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/NodeMetadataSocialProofGenerator.java

```diff
@@ -0,0 +1,182 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package com.twitter.graphjet.algorithms.socialproof;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Random;
+
+import com.twitter.graphjet.algorithms.IDMask;
+import com.twitter.graphjet.algorithms.RecommendationAlgorithm;
+import com.twitter.graphjet.algorithms.RecommendationInfo;
+import com.twitter.graphjet.algorithms.RecommendationType;
+import com.twitter.graphjet.algorithms.TweetIDMask;
+import com.twitter.graphjet.bipartite.NodeMetadataLeftIndexedPowerLawMultiSegmentBipartiteGraph;
+import com.twitter.graphjet.bipartite.NodeMetadataMultiSegmentIterator;
+import com.twitter.graphjet.hashing.IntArrayIterator;
+
+import it.unimi.dsi.fastutil.bytes.*;
+import it.unimi.dsi.fastutil.ints.*;
+import it.unimi.dsi.fastutil.longs.*;
+
+/**
+ * NodeMetadataSocialProofGenerator shares similar logic with
+ * {@link com.twitter.graphjet.algorithms.counting.TopSecondDegreeByCount}.
+ * In the request, clients specify a seed user set (left nodes) and a node metadata set
+ * (right node's metadata).
+ * NodeMetadataSocialProofGenerator finds the intersection between the seed users (left nodes)
+ * and the given node metadata id set. It accomplishes this by traversing each outgoing edge from
+ * the seed set and each right node's metadata from those edges.
+ * Only node metadatas with at least one social proof will be returned to clients.
+ */
+public abstract class NodeMetadataSocialProofGenerator implements
+  RecommendationAlgorithm<NodeMetadataSocialProofRequest, SocialProofResponse> {
+
+  private static final int MAX_EDGES_PER_NODE = 500;
+  private static final Byte2ObjectMap<Long2ObjectMap<LongSet>> EMPTY_SOCIALPROOF_MAP =
+    new Byte2ObjectArrayMap<>();
+
+  private NodeMetadataLeftIndexedPowerLawMultiSegmentBipartiteGraph graph;
+  // NodeMetadata (Int) -> Engagement (Byte) -> User (Long) -> Tweets (LongSet)
+  private final Int2ObjectMap<Byte2ObjectMap<Long2ObjectMap<LongSet>>> socialProofs;
+  // NodeMetadata (Int) -> Sum of social proof edges (Double)
+  private final Int2DoubleMap socialProofWeights;
+  protected RecommendationType recommendationType;
+  protected IDMask idMask;
+
+  public NodeMetadataSocialProofGenerator(
+    NodeMetadataLeftIndexedPowerLawMultiSegmentBipartiteGraph graph,
+    TweetIDMask mask,
+    RecommendationType recommendationType
+  ) {
+    this.graph = graph;
+    this.idMask = mask;
+    this.recommendationType = recommendationType;
+    // Variables socialProofs and socialProofWeights are re-used for each request.
+    // We chose to use an ArrayMap since, on average, we will request social proof for very few
+    // (<10) metadatas in a single request.
```

Comment:
this assumption (<10) is just from our first use case, and once we expose the api, there will be clients sending requests more than 10 to this api. it is better not to rely on the mercy of clients in library implementation. e.g. we can choose OpenHashSetMap here to future proof the potential performance degradation, and the memory is re-used anyway so we are less concerned about the memory overhead of OpenHashMap compared with ArrayMap